src/compiler/ir/README.md

```diff
@@ -0,0 +1,162 @@
+# General structure
+
+The IR has a fairly standard structure, made up of basic blocks. Each basic
+block has a list of statement and a terminator. While the implementation
+occasionally refers to "phi nodes", the IR uses basic block arguments rather
+than phi nodes.
+
+The IR only represents the body of methods. The "shape" of the program, eg. its
+classes, fields, method names and signatures, ... are all stored in the AST
+instead. For each `Method` node in the AST, a corresponding `FnAnalysis` is
+created, which, among others, contains the `MethodIR`.
+
+A `MethodIR` is composed of one or more `FunctionIR`. This distinction is used
+to encode the bodies of `when` clauses that appear within the method. The first
+`FunctionIR` corresponds to the method body itself, whereas subsequent blocks
+correspond to when block bodies.
+
+Once created by the `ir/builder` pass, the IR is immutable. All new information
+is represented in external side structures, as part of the `FnAnalysis`
+structure.
+
+# Type information
+
+The IR itself contains almost no type information. Only types which appear in
+the method body at the source level is included in the IR. At the moment, this
+is just types used in a match block.
+
+The inferred type of IR values is stored in maps in `FnAnalysis`. The type of
+`v` in basic block `bb` can be found in `analysis->typecheck->types[bb][v]`.
+
+Similarily, at construction, every method call site is assigned a
+`TypeArgumentsId`, used to identify the call's type arguments. Given an `id`,
+the inferred type arguments can be found in
+`analysis->typecheck->type_arguments[id]`.
+
+In both cases however, these are the polymorphic types. If a method or class is
+generic, these types may mention type variables. The non-generic types will
+only be known after reachabilitiy and reification.
+
+# Statements
+## `NewStmt`
+
+Allocates a new object of a given class. It may optionally be given a `parent`
+which designates the region in which the object should be allocated. If no
+parent is specified, the object is allocated in a new region.
+
+This will correspond to either a `rt::RegionTrace::create` or
+`rt::RegionTrace::alloc` call, dependending on whether a new region needs to be
+allocated.
+
+The `parent` specified in the statement could be any object in the target
+region; it may not be a reference to the region's entrypoint. On the other
+hand, the runtime's `RegionTrace::alloc` method expects to be given the
+region's entrypoint. The compiler therefore needs to do some work to find the
+entrypoint given the specified parent.
```

Comment:
When allocating an object you always need the region, because there's a bunch of metadata the runtime needs to update.

What the VM does is that every object has a pointer to its region. Whenever we want to get the region of a reference, we look up this pointer. Because regions can be merged, and we don't want to update every object to point to the new region, [this is actually a union-find](https://github.com/microsoft/verona/blob/322dc03dfbfcbf34a2de3d47ae337cdfe07fa88d/src/interpreter/object.cc#L50-L61), which is very similar to your tree idea.

However this wastes 8 bytes on every object, so it's very probably not what we want to do in the real backend, and instead have "shadow" references for every reference (and hope copy elision removes most of these). We could still do the extra pointer in the object as an initial version.

I think @mjp41 was planning on writing a more detailed section about this.